boutdata/collect.py

```diff
@@ -307,392 +236,776 @@ def getDataFile(i):
-            # one-element array and has been read as a scalar.
-            nt = 1
-        if tind_auto:
-            for i in range(nfiles):
-                t_array_ = getDataFile(i).read("t_array")
-                nt = min(len(t_array_), nt)
+        is_fieldperp = False
 
+    for i in range(grid_info["npes"]):
+        f = getDataFile(i)
+        temp_yindex, temp_f_attributes = _collect_from_one_proc(
+            i,
+            f,
+            varname,
+            result=data,
+            is_fieldperp=is_fieldperp,
+            dimensions=dimensions,
+            grid_info=grid_info,
+            tind=tind,
+            xind=xind,
+            yind=yind,
+            zind=zind,
+            xguards=xguards,
+            yguards=(yguards is not False),
+            info=info,
+        )
+        if is_fieldperp:
+            (
+                yindex_global,
+                fieldperp_yproc,
+                var_attributes,
+            ) = _check_fieldperp_attributes(
+                varname,
+                yindex_global,
+                temp_yindex,
+                i // grid_info["nxpe"],
+                fieldperp_yproc,
+                var_attributes,
+                temp_f_attributes,
+            )
+        if datafile_cache is None:
+            # close the DataFile if we are not keeping it in a cache
+            f.close()
+
+    # if a step was requested in x or y, need to apply it here
+    if xind.step is not None or yind.step is not None:
+        if dimensions == ("t", "x", "y", "z"):
+            data = data[:, :: xind.step, :: yind.step]
+        elif dimensions == ("x", "y", "z"):
+            data = data[:: xind.step, :: yind.step, :]
+        elif dimensions == ("t", "x", "y"):
+            data = data[:, :: xind.step, :: yind.step]
+        elif dimensions == ("t", "x", "z"):
+            data = data[:, :: xind.step, :]
+        elif dimensions == ("x", "y"):
+            data = data[:: xind.step, :: yind.step]
+        elif dimensions == ("x", "z"):
+            data = data[:: xind.step, :]
+        else:
+            raise ValueError(
+                "Incorrect dimensions " + str(dimensions) + " applying steps in collect"
+            )
+
+    # Finished looping over all files
     if info:
-        print("mxsub = %d mysub = %d mz = %d\n" % (mxsub, mysub, mz))
+        sys.stdout.write("\n")
+    return BoutArray(data, attributes=var_attributes)
 
-    # Get the version of BOUT++ (should be > 0.6 for NetCDF anyway)
-    try:
-        version = f["BOUT_VERSION"]
-    except KeyError:
-        print("BOUT++ version : Pre-0.2")
-        version = 0
-    if version < 3.5:
-        # Remove extra point
-        nz = mz-1
-    else:
-        nz = mz
 
-    # Fallback to sensible (?) defaults
+def _collect_from_single_file(
+    f,
+    varname,
+    xind,
+    yind,
+    zind,
+    tind,
+    path,
+    yguards,
+    xguards,
+    info,
+    prefix,
+    strict,
+    datafile_cache,
```

Comment:
I don't think so, because we actually use all of these variables within the `_collect_from_single_file`, so if we got them as `**kwargs` we'd have to unpack the `kwargs` dict, or refer to them as `kwargs["varname"]`, etc.